src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs

```diff
@@ -33,55 +31,84 @@ internal abstract partial class AbstractNavigateToSearchService
+                var nameMatches = ArrayBuilder<PatternMatch>.GetInstance();
+                var containerMatches = ArrayBuilder<PatternMatch>.GetInstance();
+
+                try
                 {
-                    if (searchDocument != null && document != searchDocument)
-                    {
-                        continue;
-                    }
+                    return await FindNavigableDeclaredSymbolInfosAsync(
+                        project, searchDocument, nameMatcher, containerMatcher,
+                        nameMatches, containerMatches, cancellationToken).ConfigureAwait(false);
+                }
+                finally
+                {
+                    nameMatches.Free();
+                    containerMatches.Free();
+                }
+            }
+        }
+
+        private static async Task<ImmutableArray<INavigateToSearchResult>> FindNavigableDeclaredSymbolInfosAsync(
+            Project project, Document searchDocument,
+            PatternMatcher nameMatcher, PatternMatcher containerMatcher,
+            ArrayBuilder<PatternMatch> nameMatches, ArrayBuilder<PatternMatch> containerMatches,
+            CancellationToken cancellationToken)
+        {
+            var result = ArrayBuilder<INavigateToSearchResult>.GetInstance();
+            foreach (var document in project.Documents)
+            {
+                if (searchDocument != null && document != searchDocument)
+                {
+                    continue;
+                }
+
+                cancellationToken.ThrowIfCancellationRequested();
+                var declarationInfo = await document.GetSyntaxTreeIndexAsync(cancellationToken).ConfigureAwait(false);
+
+                foreach (var declaredSymbolInfo in declarationInfo.DeclaredSymbolInfos)
+                {
+                    nameMatches.Clear();
+                    containerMatches.Clear();
 
                     cancellationToken.ThrowIfCancellationRequested();
-                    var declarationInfo = await document.GetSyntaxTreeIndexAsync(cancellationToken).ConfigureAwait(false);
-
-                    foreach (var declaredSymbolInfo in declarationInfo.DeclaredSymbolInfos)
-                    {
-                        cancellationToken.ThrowIfCancellationRequested();
-                        var patternMatches = patternMatcher.GetMatches(
-                            declaredSymbolInfo.Name,
-                            declaredSymbolInfo.FullyQualifiedContainerName,
-                            includeMatchSpans: true);
-
-                        if (!patternMatches.IsEmpty)
-                        {
-                            result.Add(ConvertResult(containsDots, declaredSymbolInfo, document, patternMatches));
-                        }
+                    if (nameMatcher.AddMatches(declaredSymbolInfo.Name, nameMatches) &&
+                        containerMatcher?.AddMatches(declaredSymbolInfo.FullyQualifiedContainerName, containerMatches) != false)
```

Comment:
i'll mark it 'opt'.  The usage of dot is on the matches scratch-buffer, not the matcher itself.